File: 3.x.x/crates-utopia/span/src/exporter/pretty.rs
use std::io::Write;

use crate::attr::AttrValue;
use crate::error::SpanError;
use crate::exporter::Exporter;
use crate::span::Span;

const RESET: &str = "\x1b[0m";
const BOLD: &str = "\x1b[1m";
const DIM: &str = "\x1b[2m";
const RED: &str = "\x1b[31m";
const GREEN: &str = "\x1b[32m";
const YELLOW: &str = "\x1b[33m";
const CYAN: &str = "\x1b[36m";
const WHITE: &str = "\x1b[37m";

type Sampler = Box<dyn Fn(&Span) -> bool + Send + Sync>;

/// Colourful human-readable exporter (PHP `Exporter\Pretty`).
pub struct Pretty {
    sampler: Sampler,
    max_trace_frames: usize,
    width: usize,
}

impl std::fmt::Debug for Pretty {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Pretty")
            .field("max_trace_frames", &self.max_trace_frames)
            .field("width", &self.width)
            .finish_non_exhaustive()
    }
}

impl Pretty {
    pub fn new() -> Self {
        Self::new_with(None, 3, 60)
    }

    pub fn new_with(sampler: Option<Sampler>, max_trace_frames: usize, width: usize) -> Self {
        Self {
            sampler: sampler.unwrap_or_else(|| Box::new(|_| true)),
            max_trace_frames,
            width,
        }
    }

    pub fn format(&self, span: &Span) -> String {
        let error = span.get_error();
        let has_error = error.is_some();
        let mut lines = Vec::new();
        lines.push(self.header(span, has_error));
        lines.push(String::new());

        let mut attributes: Vec<(String, AttrValue)> = span
            .get_attributes()
            .into_iter()
            .filter(|(key, _)| !key.starts_with("span."))
            .collect();

        if let Some(idx) = attributes.iter().position(|(k, _)| k == "level") {
            let level = attributes.remove(idx);
            attributes.insert(0, level);
        }

        let max_key_len = attributes.iter().map(|(k, _)| k.len()).max().unwrap_or(0);

        for (key, value) in &attributes {
            lines.push(attribute_line(key, value, max_key_len));
        }

        if let Some(error) = &error {
            if !attributes.is_empty() {
                lines.push(String::new());
            }
            lines.push(error_block(error));
            let limited = error.frames.iter().take(self.max_trace_frames);
            for frame in limited {
                let file = frame.file.as_deref().unwrap_or("unknown");
                let line = frame
                    .line
                    .map_or_else(|| "?".to_string(), |n| n.to_string());
                lines.push(format!("{DIM}    at {file}:{line}{RESET}"));
            }
            if error.frames.len() > self.max_trace_frames {
                let remaining = error.frames.len() - self.max_trace_frames;
                lines.push(format!("{DIM}    ... {remaining} more{RESET}"));
            }
        }

        lines.push(String::new());
        lines.push(format!("{DIM}{}{RESET}", "─".repeat(self.width)));
        lines.join("\n")
    }

    fn header(&self, span: &Span, has_error: bool) -> String {
        let action_color = if has_error { RED } else { GREEN };
        let action_str = format!("{BOLD}{action_color}{}{RESET}", span.get_action());
        let mut parts = vec![action_str];
        if let Some(duration) = span.get("span.duration").and_then(|v| v.as_f64()) {
            let duration_str = format_duration(duration);
            let duration_color = duration_color(duration);
            parts.push(format!("{duration_color}{duration_str}{RESET}"));
        }
        if let Some(trace_id) = span
            .get("span.trace_id")
            .and_then(|v| v.as_str().map(str::to_string))
        {
            let short: String = trace_id.chars().take(8).collect();
            parts.push(format!("{DIM}{short}{RESET}"));
        }
        parts.join(&format!("{DIM} · {RESET}"))
    }
}

impl Default for Pretty {
    fn default() -> Self {
        Self::new()
    }
}

impl Exporter for Pretty {
    fn sample(&self, span: &Span) -> bool {
        (self.sampler)(span)
    }

    fn export(&self, span: &Span) {
        let output = self.format(span);
        let stream: &mut dyn Write = if span.get_error().is_some() {
            &mut std::io::stderr()
        } else {
            &mut std::io::stdout()
        };
        let _ = writeln!(stream, "{output}");
    }
}

fn attribute_line(key: &str, value: &AttrValue, pad_to: usize) -> String {
    let padded = format!("{key:<pad_to$}");
    format!("  {CYAN}{padded}{RESET} {WHITE}{}{RESET}", value.display())
}

fn error_block(error: &SpanError) -> String {
    format!(
        "{RED}{BOLD}  {}{RESET}{RED}: {}{RESET}\n{DIM}    at {}:{}{RESET}",
        error.type_name, error.message, error.file, error.line
    )
}

fn format_duration(seconds: f64) -> String {
    if seconds >= 1.0 {
        format!("{}s", php_round(seconds, 2))
    } else {
        format!("{}ms", php_round(seconds * 1000.0, 1))
    }
}

fn duration_color(seconds: f64) -> &'static str {
    if seconds >= 1.0 {
        RED
    } else if seconds >= 0.1 {
        YELLOW
    } else {
        GREEN
    }
}

fn php_round(value: f64, digits: i32) -> f64 {
    let factor = 10f64.powi(digits);
    (value * factor).round() / factor
}

File: packages/span/src/Span/Exporter/Pretty.php
<?php

namespace Utopia\Span\Exporter;

use Closure;
use Utopia\Span\Span;

/**
 * Exports spans as colourful, human-readable multi-line output.
 *
 * Designed for local development. Writes error spans to stderr
 * and non-error spans to stdout.
 */
readonly class Pretty implements Exporter
{
    private const string RESET = "\033[0m";
    private const string BOLD = "\033[1m";
    private const string DIM = "\033[2m";

    private const string RED = "\033[31m";
    private const string GREEN = "\033[32m";
    private const string YELLOW = "\033[33m";
    private const string CYAN = "\033[36m";
    private const string WHITE = "\033[37m";

    /** @var Closure(Span): bool */
    private Closure $sampler;

    /**
     * @param Closure(Span): bool|null $sampler Filter function. Defaults to exporting every span.
     * @param int $maxTraceFrames Maximum stacktrace frames to include for errors
     * @param int $width Line width for the separator
     */
    public function __construct(
        ?Closure $sampler = null,
        private int $maxTraceFrames = 3,
        private int $width = 60,
    ) {
        $this->sampler = $sampler ?? static fn(Span $span): bool => true;
    }

    public function sample(Span $span): bool
    {
        return ($this->sampler)($span);
    }

    public function export(Span $span): void
    {
        $error = $span->getError();
        $hasError = $error instanceof \Throwable;
        $stream = $hasError ? STDERR : STDOUT;

        $lines = [];

        $lines[] = $this->header($span, $hasError);
        $lines[] = '';

        $attributes = [];
        foreach ($span->getAttributes() as $key => $value) {
            if (!str_starts_with($key, 'span.')) {
                $attributes[$key] = $value;
            }
        }

        if (\array_key_exists('level', $attributes)) {
            $level = $attributes['level'];
            unset($attributes['level']);
            $attributes = ['level' => $level] + $attributes;
        }

        $maxKeyLen = 0;
        foreach (array_keys($attributes) as $key) {
            $maxKeyLen = max($maxKeyLen, \strlen($key));
        }

        foreach ($attributes as $key => $value) {
            $lines[] = $this->attribute($key, $value, $maxKeyLen);
        }

        if ($hasError) {
            if (\count($attributes) > 0) {
                $lines[] = '';
            }

            $lines[] = $this->error($error);

            $trace = $error->getTrace();
            $limited = \array_slice($trace, 0, $this->maxTraceFrames);

            foreach ($limited as $frame) {
                $file = $frame['file'] ?? 'unknown';
                $line = $frame['line'] ?? '?';
                $lines[] = self::DIM . "    at {$file}:{$line}" . self::RESET;
            }

            $remaining = \count($trace) - $this->maxTraceFrames;
            if ($remaining > 0) {
                $lines[] = self::DIM . "    ... {$remaining} more" . self::RESET;
            }
        }

        $lines[] = '';
        $lines[] = self::DIM . str_repeat('─', $this->width) . self::RESET;

        fwrite($stream, implode(PHP_EOL, $lines) . PHP_EOL);
    }

    private function header(Span $span, bool $hasError): string
    {
        $action = $span->getAction();
        $duration = $span->get('span.duration');
        $traceId = $span->get('span.trace_id');

        $actionColor = $hasError ? self::RED : self::GREEN;
        $actionStr = self::BOLD . $actionColor . $action . self::RESET;

        $parts = [$actionStr];

        if (\is_float($duration)) {
            $durationStr = $this->formatDuration($duration);
            $durationColor = $this->durationColor($duration);
            $parts[] = $durationColor . $durationStr . self::RESET;
        }

        if (\is_string($traceId)) {
            $short = substr($traceId, 0, 8);
            $parts[] = self::DIM . $short . self::RESET;
        }

        return implode(self::DIM . ' · ' . self::RESET, $parts);
    }

    private function attribute(string $key, string|int|float|bool|null $value, int $padTo): string
    {
        $padded = str_pad($key, $padTo);
        $keyStr = self::CYAN . $padded . self::RESET;
        $valStr = self::WHITE . $this->formatValue($value) . self::RESET;

        return "  {$keyStr} {$valStr}";
    }

    private function error(\Throwable $error): string
    {
        $type = $error::class;
        $message = $error->getMessage();
        $file = $error->getFile();
        $line = $error->getLine();

        return self::RED . self::BOLD . "  {$type}" . self::RESET
            . self::RED . ": {$message}" . self::RESET . PHP_EOL
            . self::DIM . "    at {$file}:{$line}" . self::RESET;
    }

    private function formatValue(string|int|float|bool|null $value): string
    {
        return match (true) {
            $value === null => 'null',
            $value === true => 'true',
            $value === false => 'false',
            default => (string) $value,
        };
    }

    private function formatDuration(float $seconds): string
    {
        if ($seconds >= 1.0) {
            return round($seconds, 2) . 's';
        }

        return round($seconds * 1000, 1) . 'ms';
    }

    private function durationColor(float $seconds): string
    {
        if ($seconds >= 1.0) {
            return self::RED;
        }

        if ($seconds >= 0.1) {
            return self::YELLOW;
        }

        return self::GREEN;
    }
}
